src/main/java/com/google/error_analyzer/backend/StackTrace.java

```diff
@@ -0,0 +1,157 @@
+                    stackLogLines.addAll(msgsBeforeStack);
+                }
+                logger.info("Adding to stack list log line number "
+                .concat(logLineNumber.toString()));
+                storedInStack++;
+                stackLogLines.add(logText);
+            } else {
+                if (storedInStack == 0 && countOfMsgsBeforeStack < ALLOWED_MESSAGES) {
+                    //temporary storage of log lines till stack trace is found
+                    logger.info("Adding ".concat(logLineNumber.toString())
+                    .concat(" to error msgs"));
+                    msgsBeforeStack.add(logText);
+                    countOfMsgsBeforeStack++;
+                } else {
+                    //if no stack is  found or end of stack is found
+                    break;
+                }
+            }
+        }
+
+        if (storedInStack == 0) {
+            return noStackFound();
+        }
+        //reaches the end of requested documents 
+        if (storedInStack == rangeHits.length) {
+            stackLogLines.addAll(
+            findStackExceedingBatchSize(lastCheckedLine, fileName));
+        }
+        return stackLogLines.build();
+    }
+
+    public ImmutableList < String > noStackFound() {
+        return ImmutableList.<String>builder() 
+            .add("No stack found for this error").build(); 
+    }
+
+    private ImmutableList < String > findStackExceedingBatchSize(Integer errorLogLineNumber,
+    String fileName) throws IOException {
+        //Stack exceeded the batch request size
+        Builder < String > stackLogLines = ImmutableList.< String > builder();
+        Integer lastCheckedLine = errorLogLineNumber;
+        while (true) {
+            logger.info("Exceeding batch size request. Creating a new search request");
+            SearchRequest searchRequest = createSearchRequest(fileName, lastCheckedLine);
+            SearchHit[] rangeHits = logDao.getHitsFromIndex(searchRequest);
+            if (rangeHits.length == 0) {
+                return stackLogLines.build();
+            }
+            for (SearchHit rangeHit : rangeHits) {
+                Map < String ,Object > sourceMap = rangeHit.getSourceAsMap();
+                String logText = (String) sourceMap.get(LogFields.LOG_TEXT);
+                Integer logLineNumber = (Integer) sourceMap
+                .get(LogFields.LOG_LINE_NUMBER);
+                if (logLineNumber != lastCheckedLine + 1){
+                    return stackLogLines.build();
+                }
+                lastCheckedLine = logLineNumber;
+                if ( StackTraceFormat.matchesFormat(logText) ) {
+                    logger.info("Adding to stack list log line number "
+                    .concat(logLineNumber.toString()));
+                    stackLogLines.add(logText);
+                } else {
+                    return stackLogLines.build();
+                }
+            }
+        }
+    }
```

Comment:
Most of it is repetitive (similar to findStack logic). Make it modular.

Reply:
Updated